collections/unstable_binary_search.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018-2025 the Deno authors. MIT license.
+// This module is browser compatible.
+
+/**
+ * A binary search that accounts for non-exact matches.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @typeParam T The type of `haystack`.
+ *
+ * @param haystack The array to search. This MUST be sorted in ascending order, otherwise results may be incorrect.
+ * @param needle The value to search for.
+ * @returns
+ * - If `needle` is matched exactly, the index of `needle`. If multiple elements in `haystack` are equal to `needle`,
+ *   the index of the first match found (which may not be the first sequentially) is returned.
+ * - Otherwise, the [bitwise complement](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_NOT)
+ *   of `needle`'s insertion index if it were added to `haystack` in sorted order.
+ *
+ * Return value semantics are the same as C#'s [`Array.BinarySearch`](https://learn.microsoft.com/en-us/dotnet/api/system.array.binarysearch#system-array-binarysearch(system-array-system-object))
+ * and Java's [`Arrays.binarySearch`](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#binarySearch-int:A-int-).
```

Comment:
This information is useful for design decisions, but doesn't seem very useful for the users of this function. I think it is better not mentioned here.